feat: Feature/custom callout icon

lucastucious opened the pull request:
As say in #724, i wanted a simple way to add custom icons for callouts.

To handle that, i made it pure css. Icon are now a mask for the callout-icon div, so they always follow the --color form the current callout.

Now to add a custom icon, you simply add

```css
.callout {
  &[data-callout="custom"] {
    --color: #customcolor;
    --border: #custombordercolor;
    --bg: #custombg;
    --callout-icon: url('data:image/svg+xml; utf8, <custom formatted svg>');

  }
```

to custom.scss

lucastucious committed:
Add icons as masks

To handle a simple way to add custom icons, i made it pure css. Icon are now a mask for the callout-icon div, so they always follow the --color form the current callout.

Now to add a custom icon, you simply add

```css
.callout {
  &[data-callout="custom"] {
    --color: #customcolor;
    --border: #custombordercolor;
    --bg: #custombg;
    --callout-icon: url('data:image/svg+xml; utf8, <custom formatted svg>');

  }
```

to custom.scss

lucastucious committed:
remove now unused code

lucastucious commented:
Maybe i should update the doc about it ?

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -44,36 +44,20 @@ const defaultOptions: Options = {
-  quoteIcon: `<svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M3 21c3 0 7-1 7-8V5c0-1.25-.756-2.017-2-2H4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2 1 0 1 0 1 1v1c0 1-1 2-2 2s-1 .008-1 1.031V20c0 1 0 1 1 1z"></path><path d="M15 21c3 0 7-1 7-8V5c0-1.25-.75 ...
-}
-
 const callouts = {
```
now that we don't use this anymore, let's turn this into a big string literal enum instead and replace its only usage in the type of callout mapping on L63

lucastucious replied:
made an update, tell me if this what you have in mind

lucastucious committed:
Make callouts an enum

lucastucious committed:
docs: update instructions for custom callouts

jackyzha0 commented on docs/features/callouts.md:
```diff
@@ -24,7 +24,22 @@ See [documentation on supported types and syntax here](https://help.obsidian.md
 
 - Disable callouts: simply pass `callouts: false` to the plugin: `Plugin.ObsidianFlavoredMarkdown({ callouts: false })`
-- Editing icons: `quartz/plugins/transformers/ofm.ts`
+- Editing icons: `quartz\styles\callouts.scss`
```
please keep paths as forward slash `/`

jackyzha0 commented on docs/features/callouts.md:
````diff
@@ -24,7 +24,22 @@ See [documentation on supported types and syntax here](https://help.obsidian.md
+- Editing icons: `quartz\styles\callouts.scss`
+
+### Add custom callouts
+By default, custom callouts are handled by applying the "note" style. To make fancy ones, you have to add thes lines to `quartz\styles\custom.scss`
+```css
+.callout {
+  &[data-callout="custom"] {
+    --color: #customcolor;
+    --border: #custombordercolor;
+    --bg: #custombg;
+    --callout-icon: url('data:image/svg+xml; utf8, <custom formatted svg>'); //SVG icon code
+
+  }
+```
````
```suggestion
By default, custom callouts are handled by applying the `note` style. To make fancy ones, you have to add these lines to `custom.scss`.

```scss title="`quartz/styles/custom.scss"
.callout {
  &[data-callout="custom"] {
    --color: #customcolor;
    --border: #custombordercolor;
    --bg: #custombg;
    --callout-icon: url('data:image/svg+xml; utf8, <custom formatted svg>'); //SVG icon code

  }
```
```

jackyzha0 replied:
please run `npm run format`

jackyzha0 commented on quartz/styles/callouts.scss:
```diff
@@ -104,6 +128,16 @@
+  background-size: 18px 18px;
+  background-position: center;
+  background-color: var(--color);
+  -webkit-mask-image: var(--callout-icon);
```
```suggestion
```

our css engine will generate the vendor prefixes automatically, safe to remove

lucastucious replied:
That's why I was having it generated 2 times !

jackyzha0 commented on quartz/styles/callouts.scss:
```diff
@@ -104,6 +128,16 @@
   flex: 0 0 18px;
   padding-top: 4px;
+
+  //Handling icons
```
```suggestion
  // icon support
```

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -44,72 +44,42 @@ const defaultOptions: Options = {
-  quote: icons.quoteIcon,
-}
-
-const calloutMapping: Record<string, keyof typeof callouts> = {
```
we need to maintain this mapping so that aliases continue to be supported

lucastucious replied:
I have walked another path : see https://github.com/jackyzha0/quartz/pull/727/commits/195352de69939d699f67f52501bc8f5b7baca1d8.

lucastucious committed:
Prettier & run format

lucastucious committed:
dynamic matching

For maintainability, make dynamic mathching. If we or Obsidian want to support more callouts, we simply add it to the enum

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -44,72 +44,43 @@ const defaultOptions: Options = {
-  bugIcon: `<svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><rect width="8" height="14" x="8" y="6" rx="4"></rect><path d="m19 7-3 2"></path><path d="m5 7 3 2"></path><path d="m19 19-3-2"></path><path d="m5 19 3-2"></path><path d="M20 13h-4"></path><path d="M4 1 ...
-  listIcon: `<svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><line x1="8" y1="6" x2="21" y2="6"></line><line x1="8" y1="12" x2="21" y2="12"></line><line x1="8" y1="18" x2="21" y2="18"></line><line x1="3" y1="6" x2="3.01" y2="6"></line><line x1="3" y1="12" x2="3. ...
-  quoteIcon: `<svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M3 21c3 0 7-1 7-8V5c0-1.25-.756-2.017-2-2H4c-1.25 0-2 .75-2 1.972V11c0 1.25.75 2 2 2 1 0 1 0 1 1v1c0 1-1 2-2 2s-1 .008-1 1.031V20c0 1 0 1 1 1z"></path><path d="M15 21c3 0 7-1 7-8V5c0-1.25-.75 ...
+enum callouts {
```
prefer `const enum`: https://www.typescriptlang.org/docs/handbook/enums.html#const-enums

lucastucious replied:
I don't think it will work with my current approach. I'll try

lucastucious replied:
That's what i tought.

To be able to use const enum, i need to make a mapping var. 

Is this really worth it ? 

example :
```typescript
const enum callouts {
  note = "note",
  abstract = "abstract",
  summary = "abstract",
  tldr = "abstract",
  info = "info",
  todo = "todo",
  tip = "tip",
  hint = "tip",
  important = "tip",
  success = "success",
  check = "success",
  done = "success",
  question = "question",
  help = "question",
  faq = "question",
  warning = "warning",
  attention = "warning",
  caution = "warning",
  failure = "failure",
  missing = "failure",
  fail = "failure",
  danger = "danger",
  error = "danger",
  bug = "bug",
  example = "example",
  quote = "quote",
  cite = "quote",
}

const calloutsMapping: Record<string, callouts> = {
  note: callouts.note,
  abstract: callouts.abstract,
  summary: callouts.abstract,
  tldr: callouts.abstract,
  info: callouts.info,
  todo: callouts.todo,
  tip: callouts.tip,
  hint: callouts.tip,
  important: callouts.tip,
  success: callouts.success,
  check: callouts.success,
  done: callouts.success,
  question: callouts.question,
  help: callouts.question,
  faq: callouts.question,
  warning: callouts.warning,
  attention: callouts.warning,
  caution: callouts.warning,
  failure: callouts.failure,
  missing: callouts.failure,
  fail: callouts.failure,
  danger: callouts.danger,
  error: callouts.danger,
  bug: callouts.bug,
  example: callouts.example,
  quote: callouts.quote,
  cite: callouts.quote,
};

function canonicalizeCallout(calloutName: string): string {
  const normalizedCalloutName = calloutName.toLowerCase()
  if (normalizedCalloutName in calloutsMapping) {
    return (calloutsMapping[normalizedCalloutName]
  }
  // if callout is not recognized, make it a custom one
  return calloutName
}
```

lucastucious committed:
callout mapping const

Getting ride of the enum entierly as it's not worth here?

lucastucious committed:
fix callout icon styling

lucastucious committed:
Add forgotten icons

lucastucious committed:
Rebase

lucastucious commented:
As this rework how callout icons are handled, maybe the fold one should follow the same principles !

lucastucious committed:
harmonize callout icon and fold icon

lucastucious committed:
fix docs + prettier

jackyzha0 commented on docs/features/callouts.md:
````diff
@@ -24,7 +24,24 @@ See [documentation on supported types and syntax here](https://help.obsidian.md
+```
+
+> [!warning]
+> Don't forget to parse SVG code for url. You can use tools like [this one](https://yoksel.github.io/url-encoder/) F
````
```suggestion
> Don't forget to ensure that the SVG is URL encoded before putting it in the CSS. You can use tools like [this one](https://yoksel.github.io/url-encoder/)
```

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -44,39 +44,7 @@ const defaultOptions: Options = {
-}
-
-const calloutMapping: Record<string, keyof typeof callouts> = {
+const calloutMapping: Record<string, string> = {
```
```suggestion
const calloutMapping = {
```

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -106,10 +74,13 @@ const calloutMapping: Record<string, keyof typeof callouts> = {
   cite: "quote",
 }
```
```suggestion
} as const
```

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -106,10 +74,13 @@ const calloutMapping: Record<string, keyof typeof callouts> = {
-function canonicalizeCallout(calloutName: string): keyof typeof callouts {
-  let callout = calloutName.toLowerCase() as keyof typeof calloutMapping
+function canonicalizeCallout(calloutName: string): keyof typeof calloutMapping {
+  const normalizedCalloutName = calloutName.toLowerCase()
+  if (normalizedCalloutName in calloutMapping) {
+    return calloutMapping[normalizedCalloutName]
+  }
   // if callout is not recognized, make it a custom one
-  return calloutMapping[callout] ?? calloutName
+  return calloutName
 }
```
can we use the old version of this function?

jackyzha0 commented on quartz/plugins/transformers/ofm.ts:
```diff
@@ -432,16 +401,14 @@ export const ObsidianFlavoredMarkdown: QuartzTransformerPlugin<Partial<Options>
-                const toggleIcon = `<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="fold">
-                  <polyline points="6 9 12 15 18 9"></polyline>
-                </svg>`
+                const toggleIcon = `<div class="fold"></div>`
```
```suggestion
                const toggleIcon = `<div class="fold-callout-icon"></div>`
```

descriptive class names!

jackyzha0 commented on quartz/styles/callouts.scss:
```diff
@@ -82,30 +119,47 @@
+  & .fold {
+    width: var(--icon-size);
+    height: var(--icon-size);
+    //flex: 0 0 18px;
```
remove extra `flex` here?

jackyzha0 commented on quartz/styles/callouts.scss:
```diff
@@ -82,30 +119,47 @@
 
 .callout-title {
   display: flex;
+  flex-direction: row;
+  flex-wrap: wrap;
+  align-content: center;
+  align-items: center;
```
are these strictly necessary?

jackyzha0 reviewed (approved):
thank you! mostly minor naming nits now, ready to merge after addressed

lucastucious committed:
Update docs/features/callouts.md

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

lucastucious committed:
Update quartz/plugins/transformers/ofm.ts

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

lucastucious committed:
Suggestions fix

lucastucious committed:
remove unecessary rules

lucastucious committed:
comment is always nice

jackyzha0 commented on docs/features/callouts.md:
````diff
@@ -24,7 +24,24 @@ See [documentation on supported types and syntax here](https://help.obsidian.md
+```
+
+> [!warning]
+> Don't forget to ensure that the SVG is URL encoded before putting it in the CSS. You can use tools like [this one](https://yoksel.github.io/url-encoder/)
````
```suggestion
> Don't forget to ensure that the SVG is URL encoded before putting it in the CSS. You can use tools like [this one](https://yoksel.github.io/url-encoder/) to help you do that.
```

jackyzha0 committed:
Update docs/features/callouts.md

jackyzha0 commented:
this is great, thanks :)
